version: "3.9"

x-common-env: &common-env
  HOST: 158.109.9.60
  BROKER_HOST: 158.109.9.60
  BROKER_PORT: 1026
  TOOLBOX_DATA: /etc/communicity_toolbox

services:
  image_storage:
    build: .
    restart: always
    environment:
      <<: *common-env
      PORT: 9001
      IMG_STORAGE_PATH: /var/communicity_toolbox/image_storage
    ports:
      - "9001:8080"
    command: python ./toolbox/Projects/ImageStorage/api.py --config /etc/communicity_toolbox/configs/ImageStorage.yaml
    volumes:
      - image_storage:/var/communicity_toolbox/image_storage
      - ./data:/etc/communicity_toolbox/:ro
  
  face_recognition_dataset_creation:
    build: .
    environment:
      <<: *common-env
    command: sudo -E python ./toolbox/Projects/FaceRecognition/dataset_creator.py --config /etc/communicity_toolbox/configs/FaceRecognition.yaml -i /etc/communicity_toolbox/samples/images/faces/celeb_single -o /etc/communicity_toolbox/samples/face_recognition_datasets/facenet_celeb_single.pkl
    volumes:
      - ./data:/etc/communicity_toolbox/:rw

  face_recognition:
    build: .
    restart: always
    depends_on:
      face_recognition_dataset_creation:
        condition: service_completed_successfully
    environment:
      <<: *common-env
      PORT: 9002
    ports:
      - "9002:8080"
    command: python ./toolbox/Projects/FaceRecognition/api.py --config /etc/communicity_toolbox/configs/FaceRecognition.yaml
    volumes:
      - image_storage:/var/communicity_toolbox/image_storage
      - ./data:/etc/communicity_toolbox/:ro
  
  age_gender:
    build: .
    restart: always
    environment:
      <<: *common-env
      PORT: 9003
    ports:
      - "9003:8080"
    command: python ./toolbox/Projects/AgeGender/api.py --config /etc/communicity_toolbox/configs/AgeGender.yaml
    volumes:
      - image_storage:/var/communicity_toolbox/image_storage
      - ./data:/etc/communicity_toolbox/:ro
  
  face_emotions:
    build: .
    restart: always
    environment:
      <<: *common-env
      PORT: 9004
    ports:
      - "9004:8080"
    command: python ./toolbox/Projects/FaceEmotions/api.py --config /etc/communicity_toolbox/configs/FaceEmotions.yaml
    volumes:
      - image_storage:/var/communicity_toolbox/image_storage
      - ./data:/etc/communicity_toolbox/:ro
  
  instance_segmentation:
    build: .
    restart: always
    environment:
      <<: *common-env
      PORT: 9005
    ports:
      - "9005:8080"
    command: python ./toolbox/Projects/InstanceSegmentation/api.py --config /etc/communicity_toolbox/configs/InstanceSegmentation.yaml
    volumes:
      - image_storage:/var/communicity_toolbox/image_storage
      - ./data:/etc/communicity_toolbox/:ro
  
  keypoints:
    build: .
    restart: always
    environment:
      <<: *common-env
      PORT: 9006
    ports:
      - "9006:8080"
    command: python ./toolbox/Projects/Keypoints/api.py --config /etc/communicity_toolbox/configs/Keypoints.yaml
    volumes:
      - image_storage:/var/communicity_toolbox/image_storage
      - ./data:/etc/communicity_toolbox/:ro

  face_detection:
    build: .
    restart: always
    environment:
      <<: *common-env
      PORT: 9007
    ports:
      - "9007:8080"
    command: python ./toolbox/Projects/FaceDetection/api.py --config /etc/communicity_toolbox/configs/FaceDetection.yaml
    volumes:
      - image_storage:/var/communicity_toolbox/image_storage
      - ./data:/etc/communicity_toolbox/:ro

  # Orion context broker
  orion:
    image: fiware/orion-ld:1.1.0
    hostname: orion
    restart: always
    depends_on:
      - mongo-db
    expose:
      - "1026"
    ports:
      - "1026:1026" 
    command: -dbhost mongo-db -logLevel DEBUG
    healthcheck:
      test: curl --fail -s http://orion:1026/version || exit 1

  # Context broker databases
  mongo-db:
    image: mongo:3.6
    hostname: mongo-db
    expose:
      - "27017"
    ports:
      - "27017:27017" 
    command: --nojournal
    volumes:
      - mongo-db:/data

volumes:
  image_storage:
  mongo-db:
